Add ShoppingCart model

ipantazi opened the pull request:
Added entity ShoppingCart and CartItem, and DTO classes: CartItemDto, CartItemSimpleDto, CreateCartItemDto, ShoppingCartDto, UpdateCartItemDto.

Added interfaces CartItemMapper, CartItemRepository, ShoppingCartMapper, ShoppingCartRepository, ShoppingCartService.

Added changes to Liquibase.

Deleted method toBookDtoWithoutCategories() to BookMapper interface, handleGenericException() and handleUserNotFoundException() methods from CustomGlobalExceptionHandler class, findByName() from RoleRepository interface, deleted AuthenticationException and UserEmailNotFoundException classes.

Added ShoppingCartServiceImpl, ShoppingCartController, UsernameNotFoundException classes.

Added annotation @EqualsAndHashCode(onlyExplicitlyIncluded = true) to User, Category, Book classes, and deleted @Setter from Role classes.

Added methods findWithCategoriesById() to BookRepository interface, findAllUsersWithoutShoppingCart() to UserRepository interface.

Added shopping cart creating to the user registration flow and deleting cart items with book.

ipantazi committed:
Added entity ShoppingCart and CartItem, and DTO classes: CartItemDto, CartItemSimpleDto, CreateCartItemDto, ShoppingCartDto, UpdateCartItemDto.

Added interfaces CartItemMapper, CartItemRepository, ShoppingCartMapper, ShoppingCartRepository, ShoppingCartService.

Added changes to Liquibase.

Deleted method toBookDtoWithoutCategories() to BookMapper interface, handleGenericException() and handleUserNotFoundException() methods from CustomGlobalExceptionHandler class, findByName() from RoleRepository interface, deleted AuthenticationException and UserEmailNotFoundException classes.

Added ShoppingCartServiceImpl, ShoppingCartController, UsernameNotFoundException classes.

Added annotation @EqualsAndHashCode(onlyExplicitlyIncluded = true) to User, Category, Book classes, and deleted @Setter from Role classes.

Added methods findWithCategoriesById() to BookRepository interface, findAllUsersWithoutShoppingCart() to UserRepository interface.

Added shopping cart creating to the user registration flow and deleting cart items with book.

Elena-Bruyako commented on src/main/java/onlinebookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,69 @@
+    @ResponseStatus(value = HttpStatus.CREATED)
+    @Operation(summary = "Add cart item to the shopping cart.",
+            description = "Add a book to the shopping cart.")
+    public CartItemDto addCartItem(Authentication authentication,
```
```suggestion
    public ShoppingCartDto addCartItem(Authentication authentication,
```

Elena-Bruyako commented on src/main/java/onlinebookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,69 @@
+    @ResponseStatus(value = HttpStatus.OK)
+    @Operation(summary = "Update cart item of a book in the shopping cart.",
+            description = "Update the books quantity in the shopping cart by cart item id.")
+    public CartItemSimpleDto updateCartItem(@PathVariable Long cartItemId,
```
```suggestion
    public ShoppingCartDto updateCartItem(Authentication authentication, @PathVariable Long cartItemId,
```

Elena-Bruyako commented on src/main/java/onlinebookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,69 @@
+
+    @PutMapping("/items/{cartItemId}")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @ResponseStatus(value = HttpStatus.OK)
```
HttpStatus OK by default, remove

Elena-Bruyako commented on src/main/java/onlinebookstore/controller/ShoppingCartController.java:
```diff
@@ -0,0 +1,69 @@
+    @ResponseStatus(value = HttpStatus.NO_CONTENT)
+    @Operation(summary = "Remove a book from the shopping cart.",
+            description = "Remove an item from the user's shopping cart.")
+    public void delete(@PathVariable Long cartItemId) {
```
```suggestion
    public void delete(Authentication authentication, @PathVariable Long cartItemId) {
```

Elena-Bruyako commented on src/main/java/onlinebookstore/dto/cartitem/CreateCartItemDto.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.Data;
+
+@Data
+public class CreateCartItemDto {
```
```suggestion
public class CartItemRequestDto {
```

Elena-Bruyako commented on src/main/java/onlinebookstore/dto/cartitem/CreateCartItemDto.java:
```diff
@@ -0,0 +1,16 @@
+    @Positive(message = "Invalid book id. Value should be positive.")
+    private Long bookId;
+
+    @NotNull(message = "Invalid quantity. Value shouldn't be null")
```
int can't be null
```suggestion
```

Elena-Bruyako commented on src/main/java/onlinebookstore/dto/cartitem/CartItemSimpleDto.java:
```diff
@@ -0,0 +1,11 @@
+
+@Data
+@AllArgsConstructor
+public class CartItemSimpleDto {
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/dto/cartitem/UpdateCartItemDto.java:
```diff
@@ -0,0 +1,15 @@
+    private Long shoppingCartId;
+    private Long bookId;
+
+    @NotNull(message = "Invalid quantity. Value shouldn't be null")
```
```suggestion
```

Elena-Bruyako commented on src/main/java/onlinebookstore/exception/UsernameNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package onlinebookstore.exception;
+
+public class UsernameNotFoundException extends RuntimeException {
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/mapper/CartItemMapper.java:
```diff
@@ -0,0 +1,30 @@
+    CartItem toCartItemEntity(CreateCartItemDto createCartItemDto);
+
+    @Mapping(target = "bookId", ignore = true)
+    @Mapping(target = "bookTitle", ignore = true)
```
```suggestion
    @Mapping(source = "book.id", target = "bookId")
    @Mapping(source = "book.title", target = "bookTitle")
```

Elena-Bruyako commented on src/main/java/onlinebookstore/mapper/CartItemMapper.java:
```diff
@@ -0,0 +1,30 @@
+    CartItemDto toCartItemDto(CartItem cartItem);
+
+    @AfterMapping
+    default void setBookInfo(@MappingTarget CartItemDto cartItemDto, CartItem cartItem) {
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/model/Role.java:
```diff
@@ -9,12 +9,10 @@
 
 @Entity
 @Table(name = "roles")
-@Setter
```
Why don't you need setter anymore?

Elena-Bruyako commented on src/main/java/onlinebookstore/model/CartItem.java:
```diff
@@ -0,0 +1,41 @@
+@Table(name = "cart_items")
+@Getter
+@Setter
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
```
I suggest not to add EqualsAndHashCode at all here and in Shop cart also

Elena-Bruyako commented on src/main/java/onlinebookstore/model/Category.java:
```diff
@@ -17,9 +18,11 @@
 @SQLRestriction("is_deleted = false")
 @Setter
 @Getter
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
```
same comment here

Elena-Bruyako commented on src/main/java/onlinebookstore/model/ShoppingCart.java:
```diff
@@ -0,0 +1,49 @@
+
+    @OneToMany(
+            mappedBy = "shoppingCart",
+            fetch = FetchType.LAZY,
```
```suggestion
```

Elena-Bruyako commented on src/main/java/onlinebookstore/repository/cartitem/CartItemRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+@Repository
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    @Modifying
+    @Query("UPDATE CartItem ci SET ci.quantity = :quantity WHERE ci.id = :cartItemId")
+    int updateQuantity(@Param("cartItemId") Long cartItemId, @Param("quantity") int quantity);
+
+    void deleteAllByBookId(Long bookId);
```
```suggestion
    Optional<CartItem> findByIdAndShoppingCartId(Long itemId, Long cartId);
```

Elena-Bruyako commented on src/main/java/onlinebookstore/repository/role/RoleRepository.java:
```diff
@@ -1,11 +1,9 @@
 
 @Repository
 public interface RoleRepository extends JpaRepository<Role, Long> {
-    Optional<Role> findByName(Role.RoleName roleName);
```
Why don't you need this method anymore?

ipantazi replied:
I don't use this method anywhere.

Elena-Bruyako commented on src/main/java/onlinebookstore/repository/shoppingcart/ShoppingCartRepository.java:
```diff
@@ -0,0 +1,25 @@
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query("""
+            SELECT DISTINCT sc FROM ShoppingCart sc
+            LEFT JOIN FETCH sc.cartItems ci
+            LEFT JOIN FETCH ci.book
+            WHERE sc.isDeleted = false AND sc.user.email = :email
+            """)
+    Optional<ShoppingCart> findByEmailAndWithoutUser(@Param("email") String email);
+
+    @EntityGraph(attributePaths = {"user", "cartItems"})
+    @NonNull
+    Optional<ShoppingCart> findByUserEmail(@NonNull String email);
```
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
    ShoppingCart findByUserId(Long userId);
```

Elena-Bruyako commented on src/main/java/onlinebookstore/repository/book/BookRepository.java:
```diff
@@ -32,8 +42,11 @@ public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificat
+
+    @Override
+    @NonNull
     Optional<Book> findById(@NonNull Long id);
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/repository/user/UserRepository.java:
```diff
@@ -1,15 +1,26 @@
 import org.springframework.stereotype.Repository;
 
 @Repository
+@EnableJpaRepositories(basePackageClasses = UserRepository.class)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/onlinebookstore/repository/user/UserRepository.java:
```diff
@@ -1,15 +1,26 @@
 
     @EntityGraph(attributePaths = "roles")
     Optional<User> findByEmail(String email);
+
+    @Query("""
+            SELECT u FROM User u
+            LEFT JOIN ShoppingCart sc ON u.id = sc.id
+            WHERE u.isDeleted = false AND sc.id IS NULL
+            """)
+    List<User> findAllUsersWithoutShoppingCart();
```
```suggestion
```

Elena-Bruyako commented on src/main/java/onlinebookstore/security/CustomUserDetailsService.java:
```diff
@@ -16,6 +15,6 @@ public class CustomUserDetailsService implements UserDetailsService {
     public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
         return userRepository.findByEmail(email).orElseThrow(()
-                -> new UserEmailNotFoundException("Can't find user by email: " + email));
+                -> new UsernameNotFoundException("Can't find user by email: " + email));
```
```suggestion
                -> new EntityNotFoundException("Can't find user by email: " + email));
```

Elena-Bruyako commented on src/main/java/onlinebookstore/security/CustomUserDetailsService.java:
```diff
@@ -16,6 +15,6 @@ public class CustomUserDetailsService implements UserDetailsService {
     @Override
     public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
```
```suggestion
    public UserDetails loadUserByUsername(String email) {
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+
+    @EventListener(ApplicationReadyEvent.class)
```
```suggestion
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    private final CartItemRepository cartItemRepository;
+
+    @EventListener(ApplicationReadyEvent.class)
+    @Transactional
```
replace under the class

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+    @EventListener(ApplicationReadyEvent.class)
+    @Transactional
+    public void initializeShoppingCart() {
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public ShoppingCartDto getShoppingCart(String email) {
+        return shoppingCartMapper.toShoppingCartDto(findShoppingCartWithoutUser(email));
```
```suggestion
        return shoppingCartMapper.toShoppingCartDto(findByUserId(userId));
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(String email) {
```
```suggestion
    public ShoppingCartDto getShoppingCart(Long userId) {
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    @Transactional
+    public CartItemDto addCartItem(String email, CreateCartItemDto createCartItemDto) {
```
don't use email for searching, it's better to use userId because it's PK (index) so searching will be faster

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    @Transactional
+    public CartItemDto addCartItem(String email, CreateCartItemDto createCartItemDto) {
+        Long bookId = createCartItemDto.getBookId();
```
Flow is

find book by id 
find shop cart by userId 
get cartItems from shop cart
if there is a book with the same id - increase quantity
if no - add to shop cart new cartItem

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+    @Override
+    @Transactional
+    public CartItemSimpleDto updateCartItem(Long cartItemId, UpdateCartItemDto updateCartItem) {
+        int updatedRows = cartItemRepository.updateQuantity(cartItemId,
+                updateCartItem.getQuantity());
+        if (updatedRows == 0) {
+            throw new EntityNotFoundException("Can't find cart item with id: " + cartItemId);
+        }
+        return new CartItemSimpleDto(cartItemId, updateCartItem.getQuantity());
```
```suggestion
        ShoppingCart cart = repository.findByUserId(userId);
        CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId()).....
        update quantity
        save cartItem
        return ShoppingCartDto
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public void deleteById(Long id) {
+        if (!cartItemRepository.existsById(id)) {
```
```
ShoppingCart shoppingCart = repository.findByUserId(userId);
CartItem cartItem = itemRepository.findByIdAndShoppingCartId(itemId, shoppingCart.getId()).....
remove cartItem from shop cart
```

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    private ShoppingCart findShoppingCartWithoutBook(String email) {
```
remove

Elena-Bruyako commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,105 @@
+                new EntityNotFoundException("Can't find shopping cart by email: " + email));
+    }
+
+    private ShoppingCart findShoppingCartWithoutUser(String email) {
```
remove

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
Removed @ResponseStatus(HttpStatus.OK) from BookController, CategoryController, ShoppingCartController classes.

Deleted method findWithCategoriesById() from BookRepository interface, findAllUsersWithoutShoppingCart() from UserRepository interface, initializeShoppingCart() and findShoppingCartWithoutBook() and findShoppingCartWithoutUser() from ShoppingCartServiceImpl class, findByEmailAndWithoutUser() and findByUserEmail() from ShoppingCartRepository interface, handleUsernameNotFoundException() from CustomGlobalExceptionHandler class, updateQuantity() and deleteAllByBookId() from CartItemRepository, setBookInfo() from interface CartItemMapper.

Added methods findShoppingCartByUserId() and findCartItemByItemIdAndCartId() to the ShoppingCartServiceImpl class, findByIdAndShoppingCartId() and findByBookId() in interface CartItemRepository, findBookById to the interface BookService.

Deleted annotation @EqualsAndHashCode(onlyExplicitlyIncluded = true) from CartItem, Category, ShoppingCart, @EnableJpaRepositories from UserRepository interface, @NotNull from CreateCartItemDto class.

Deleted fetch = FetchType.LAZY from ShoppingCart.

Deleted classes CartItemSimpleDto and UsernameNotFoundException.

Fix methods getShoppingCart(), addCartItem(), updateCartItem(), deleteById(), registerNewShoppingCart() in interface ShoppingCartService and class ShoppingCartServiceImpl, loadUserByUsername() in CustomUserDetailsService class, fix methods in class BookServiceImpl and class ShoppingCartController.

Deleted fields from class UpdateCartItemDto.

I don't use this method anywhere. - Optional<Role> findByName(Role.RoleName roleName); in interface RoleRepository.

liliia-ponomarenko commented on src/main/java/onlinebookstore/repository/book/BookRepository.java:
```diff
@@ -19,8 +20,17 @@ public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificat
-    @Query(value = "SELECT b FROM Book b JOIN b.categories c WHERE c.id = :categoryId")
-    Page<Book> findAllByCategoryId(@Param("categoryId") Long categoryId, Pageable pageable);
+    @Query("""
+            SELECT new onlinebookstore.dto.book.BookDtoWithoutCategoryIds(
```
don't return dto from repository. return book and use mapper on the service layer

liliia-ponomarenko commented on src/main/java/onlinebookstore/repository/book/BookRepository.java:
```diff
@@ -32,7 +42,6 @@ public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificat
 
-    @Override
     @EntityGraph(attributePaths = "categories")
     @NonNull
```
Remove NonNull (the same for other methods

liliia-ponomarenko commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    private final BookService bookService;
+
+    @Override
+    @Transactional
```
move transactional above class

liliia-ponomarenko commented on src/main/java/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:
```diff
@@ -0,0 +1,103 @@
+    public ShoppingCartDto addCartItem(Long userId, CartItemRequestDto cartItemRequestDto) {
+        Long bookId = cartItemRequestDto.getBookId();
+        ShoppingCart shoppingCart = findShoppingCartByUserId(userId);
+        Map<Long, CartItem> cartItemMap = shoppingCart.getCartItems().stream()
+                .collect(Collectors.toMap(
+                        cartItem -> cartItem.getBook().getId(),
+                        cartItem -> cartItem));
```
do you really need converting to the map?

liliia-ponomarenko reviewed (changes requested):
good job! A few comments ;) Please don’t mark comments as resolved. Use stickers or something  else

ipantazi committed:
Added method BookDtoWithoutCategoryIds() to BookMapper interface.

Fix method findAllByCategoryId() and removed annotation @NonNull from all methods in the BookRepository interface.

Fix methods update() and getByCategoryId() in the BookServiceImpl class.

Fix method addCartItem() and replaced annotation @Transactional in the class ShoppingCartServiceImp.

liliia-ponomarenko reviewed (approved):
Good job!
